### PR TITLE
Improve Docker startup process

### PR DESCRIPTION
**Pull request by calvinaco**
- Change Dockerfile to use pre-build image from Baidu
- Include build steps in Dockerfile
- Update README instruction

-----

<p align="center">
  Crypto.com Chain
</p>

<h2 align="center">(Work in Progress) <a href="https://crypto.com">Crypto.com<a> Chain Transaction Enclaves</h2

For more details, see the [Crypto.com Chain README](https://github.com/crypto-com/chain)

## Build the Docker image
```bash
$ docker build -t chain-tx . \
--build-arg SGX_MODE=<SW|HW> \
--build-arg NETWORK_ID=<NETWORK_HEX_ID>
```

- `SGX_MODE`:
  - `SW` for Software Simulation mode
  - `HW` for Hardware mode
- `NETWORK_HEX_ID`: Network HEX Id of Tendermint
- `APP_PORT`: Listening Port inside the Docker instance (Default: 25933)

## Run the Docker instance

- Software Simulation Mode
```bash
# docker run --rm -p <HOST_PORT>:<DOCKER_APP_PORT> -rm chain-tx
$ docker run --rm \
-p 25933:25933 \
chain-tx
```

- Hardware Mode
```bash
# docker run --rm --device /dev/isgx -p <HOST_PORT>:<DOCKER_APP_PORT> chain-tx
$ docker run --rm \
--device /dev/isgx \
-p 25933:25933 \
chain-tx
```

## Run /bin/bash inside Docker instance

If you want to get your hands dirty, you can
```bash
$ docker run --rm \
chain-tx \
/bin/bash
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,75 +1,27 @@
...
-# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
-# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
-# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
-# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
-# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
-# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-#
-# Modifications Copyright 2019 Foris Limited (licensed under the Apache License, Version 2.0)
+FROM baiduxlab/sgx-rust:1604-1.0.8
+LABEL maintainer="Crypto.com"
```
Comment (leejw51crypto) [inline Dockerfile]: hello

File: Dockerfile [hunk 2]
```diff
@@ -1,75 +1,27 @@
...
-# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
-# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
-# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
-# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
-# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
-# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
-# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-#
-# Modifications Copyright 2019 Foris Limited (licensed under the Apache License, Version 2.0)
+FROM baiduxlab/sgx-rust:1604-1.0.8
```
Comment (tomtau) [inline Dockerfile]: can it be based on 1804?

File: Dockerfile (hunk 1 above)
Comment (calvinaco) [reply Dockerfile]: Added back the license statement

File: Dockerfile (hunk 2 above)
Comment (calvinaco) [reply Dockerfile]: Let me try, it should still work

Comment (leejw51crypto) [reply Dockerfile]: make 

*** No rule to make target `/opt/intel/sgxsdk/bin/x64/sgx_edger8r', needed by `enclave/Enclave_t.c'.  Stop

Comment (leejw51crypto) [reply Dockerfile]: os is mac

Comment (calvinaco) [reply Dockerfile]: Changed to use 18.04 image @tomtau

Comment (calvinaco) [reply Dockerfile]: `/opt/intel/sgxsdk/bin/x64/sgx_edger8r`, the path prefix should not be `/opt/intel/sgxsdk`. Perhaps the `~/.docker_bashrc` is not sourced properly. We can investigate on your Mac tomorrow.

Comment (leejw51crypto) [reply Dockerfile]: ok

File: Dockerfile [hunk 3]
```diff
@@ -1,75 +1,29 @@
-# BSD License
```
Comment (tomtau) [inline Dockerfile]: If the Dockerfile has any bits from the original code, then keep it (it says it: "Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.")
If not, there's no need for the "Modifications Copyright" either

Comment (leejw51crypto) [reply Dockerfile]: in my mac, i cannot build
after this issue cleared, i will approve

File: README.md [hunk 4]
```diff
@@ -1,25 +1,50 @@
...
 
-Build:
+- Hardware Mode
+```bash
+# docker run --rm --device /dev/isgx -p <HOST_PORT>:<DOCKER_APP_PORT> chain-tx --env LD_LIBRARY_PATH="/opt/intel/libsgx-enclave-common/aesm /opt/intel/libsgx-enclave-common/aesm/aesm_service"
+$ docker run --rm \
+--device /dev/isgx \
+-p 25933:25933 \
+chain-tx \
+--env LD_LIBRARY_PATH="/opt/intel/libsgx-enclave-common/aesm /opt/intel/libsgx-enclave-common/aesm/aesm_service"
```
[the remark this answers is not in the document]
Comment (calvinaco) [reply README.md]: I have updated `entrypoint.sh` to bring up the `aesm_service` on hardware mode.

File: Dockerfile (hunk 3 above)
Comment (calvinaco) [reply Dockerfile]: I see, I have removed the statement now

**Comment by calvinaco** [comment]
I tested on the notebook with SGX and it is working fine under HW mode.

**Pull request merged**